ct: Add write_balancer component

Lazin opened the pull request:
Introduce write_request_scheduler for cross-shard write scheduling

The write_request_scheduler is a new pipeline component implementing the "parallel batcher" design. It is placed before the batcher in the pipeline and acts as a peering_sharded_service, redirecting write requests to batchers for upload.

Currently, the batcher handles uploads, aggregation, and scheduling. It decides when enough data has accumulated (N ms or M bytes) to trigger an upload. This design forces the batcher to manage multiple concerns at once. With upcoming cross-shard operations, this complexity would grow further.

In the new setup, the batcher no longer schedules writes. Instead, it simply aggregates all write requests it receives and uploads them to cloud storage. Scheduling is now handled by the write_request_scheduler across shards, using two asynchronous background fibers that implement two policies:

- **Data threshold policy** — Forces a shard to upload immediately once its local data reaches a set threshold. This operates per shard and does not cross shard boundaries.

- **Time-based fallback policy** — Runs only on shard 0. After N ms, it forces all shards to send write requests to a single target shard for upload. The target shard is chosen based on having the most data in the pipeline to minimize cross-shard data copying and improve CPU cache usage.

This separation of scheduling from batching reduces complexity, supports cross-shard coordination, and optimizes resource usage.

The batcher will be updated in the next commits. It has to be simplified. The scheduling aspect should be removed out of it.
Also, some additional notification mechanism should be implemented to avoid a situation when batcher uploads L0 object without waiting for all data sent by the scheduler. This is a simple change local to the even subscription mechanism of the pipeline.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

Lazin commented:
The plan is to pass by reference but in this draft pr the data is copied.

On Mon, Aug 18, 2025, 17:39 Travis Downs ***@***.***> wrote:

> *travisdowns* left a comment (redpanda-data/redpanda#27269)
> <https://github.com/redpanda-data/redpanda/pull/27269#issuecomment-3197211366>
>
> @Lazin <https://github.com/Lazin> wrote:
>
>
>    - minimize cross-shard data copying and improve CPU cache usage.
>
> Is there data *copying* across shards, or is it pass by reference? I
> would hope the latter: recall that iobuf allows you to have lightweight
> copies of the same data on different shards, where the underlying buffer is
> shared and reference counted.
>
> This also helps with memory accounting since you only need to
> account/limit the original allocation on the home shards, versus true
> copying where you'd need to ensure the target shard doesn't use too much
> memory.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/redpanda-data/redpanda/pull/27269#issuecomment-3197211366>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAWMNWMP6H4EQFAMYQMFU33OHQRFAVCNFSM6AAAAACEARCBG2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCOJXGIYTCMZWGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dotnwat commented:
> Is there data copying across shards, or is it pass by reference?

copying bulk data across shards is kind of hard to do by accident, right? it seems like you'd need to start by sending a reference to the destination core then on the destination core explicitly make a copy

Lazin commented:
> copying bulk data across shards is kind of hard to do by accident, right? it seems like you'd need to start by sending a reference to the destination core then on the destination core explicitly make a copy

Yes, this is what's happening. The target shard invoked function called `deep_copy`. I decided not to include this optimisation to make the PR easier to understand.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--2fde8da6-33e4-4cea-8e24-475052e58368-->
<!--TEST RESULTS-->
<!--5ba767dc-ee74-45b0-b4d7-1b2b20a9213a-->
<!--TEST RESULTS-->
<!--4e282d64-c610-4b76-b105-25c6a863079e-->
<!--TEST RESULTS-->
<!--88374219-2375-426c-9401-768820fae6ba-->
<!--TEST RESULTS-->
<!--a56c64d2-f87f-4bd9-b3d3-bc33a3726513-->
<!--TEST RESULTS-->
<!--58195c60-10f0-41b5-9228-9c559a7ff0e1-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71764">build#71764</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                        |
|:-------------------------|:---------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTest  | test_write           | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/71764#01991225-354d-49a3-b3d8-aa0e1f0a3611 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS     |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71764#01991225-354b-4dec-b099-4db13ee615b3 | FLAKY         | 20/21    | upstream reliability is '99.21875'. current run reliability is '95.23809523809523'. drift is 3.98065 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71797">build#71797</a></summary>


| test_class                                                                               | test_method                                                                              | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:-----------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| src/v/cloud_topics/level_zero/write_request_scheduler/tests/write_request_scheduler_test | src/v/cloud_topics/level_zero/write_request_scheduler/tests/write_request_scheduler_test |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/71797#0199172c-8015-4a42-b886-0f603aa3ec53 | FAIL          | 0/1      |          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71927">build#71927</a></summary>


| test_class                                                                               | test_method                                                                              | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest                                                                 | test_node_operations                                                                     | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71927#01992fa8-25f1-4bb1-8689-155e98809af0 | FLAKY         | 19/21    | upstream reliability is '98.61111111111111'. current run reliability is '90.47619047619048'. drift is 8.13492 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest                                                                 | test_node_operations                                                                     | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71927#01992fa8-25f2-47d7-b0d8-053af10de69a | FLAKY         | 20/21    | upstream reliability is '98.71244635193133'. current run reliability is '95.23809523809523'. drift is 3.47435 and the allowed drift is set to 50. The test should PASS |
| src/v/cloud_topics/level_zero/write_request_scheduler/tests/write_request_scheduler_test | src/v/cloud_topics/level_zero/write_request_scheduler/tests/write_request_scheduler_test |                                                                                                                                         | unit        | https://buildkite.com/redpanda/redpanda/builds/71927#01992f6d-5635-4522-944b-6cd23e8e6ac0 | FAIL          | 0/1      |                                                                                                                                                                        |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72024">build#72024</a></summary>


| test_class                                                                               | test_method                                                                              | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-----------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest                                                                 | test_node_operations                                                                     | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72024#0199352f-c3fd-4d84-9c8e-9faf4d241161 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
| src/v/cloud_topics/level_zero/write_request_scheduler/tests/write_request_scheduler_test | src/v/cloud_topics/level_zero/write_request_scheduler/tests/write_request_scheduler_test |                                                                                                                                                 | unit        | https://buildkite.com/redpanda/redpanda/builds/72024#019934f3-37ff-4b90-aa6b-4ac0c19156fd | FAIL          | 0/1      |                                                                                                                                                           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72222">build#72222</a></summary>


| test_class                                        | test_method                                       | test_arguments                             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:--------------------------------------------------|:--------------------------------------------------|:-------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| PandaproxyAuthSecurityReportTest                  | test_security_report                              | {"auto_auth": false, "enable_auth": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72222#01994454-57aa-490c-a85c-b7090f1a7668 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
| src/v/datalake/coordinator/tests/coordinator_test | src/v/datalake/coordinator/tests/coordinator_test |                                            | unit        | https://buildkite.com/redpanda/redpanda/builds/72222#0199441d-ee9d-45bc-937c-73c61512907e | FAIL          | 0/1      |                                                                                                                                                           |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72970">build#72970</a></summary>


| test_class              | test_method                                                     | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                                       |
|:------------------------|:----------------------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterRateQuotaTest    | test_client_response_throttle_mechanism_applies_to_next_request | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72970#019981f9-fdb2-4c4e-a5fb-9f61630dc147 | FLAKY         | 15/21    | upstream reliability is '82.3371989295272'. current run reliability is '71.42857142857143'. drift is 10.90863 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism_applies_to_next_request |
| DisablingPartitionsTest | test_disable                                                    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72970#019981f9-fdb5-43b2-967e-f6b0c0c0e60a | FLAKY         | 14/21    | upstream reliability is '82.53968253968253'. current run reliability is '66.66666666666666'. drift is 15.87302 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DisablingPartitionsTest&test_method=test_disable                                                 |
| TimeQueryTest           | test_timequery_with_local_gc                                    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72970#01998205-536c-4321-a889-3b1a5d08f48c | FLAKY         | 19/21    | upstream reliability is '97.11340206185567'. current run reliability is '90.47619047619048'. drift is 6.63721 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery_with_local_gc                                           |
</details>
<!--/body-->

Lazin committed:
ct: Add write_balancer component

Introduce write_request_scheduler for cross-shard write scheduling

The write_request_scheduler is a new pipeline component implementing the
"parallel batcher" design. It is placed before the batcher in the pipeline
and acts as a peering_sharded_service, redirecting write requests to
batchers for upload.

Currently, the batcher handles uploads, aggregation, and scheduling. It
decides when enough data has accumulated (N ms or M bytes) to trigger an
upload. This design forces the batcher to manage multiple concerns at once.
With upcoming cross-shard operations, this complexity would grow further.

In the new setup, the batcher no longer schedules writes. Instead, it simply
aggregates all write requests it receives and uploads them to cloud storage.
Scheduling is now handled by the write_request_scheduler across shards,
using two asynchronous background fibers that implement two policies:

- **Data threshold policy** — Forces a shard to upload immediately once its
local data reaches a set threshold. This operates per shard and does not
cross shard boundaries.

- **Time-based fallback policy** — Runs only on shard 0. After N ms, it
forces all shards to send write requests to a single target shard for
upload. The target shard is chosen based on having the most data in the
pipeline to minimize cross-shard data copying and improve CPU cache usage.

This separation of scheduling from batching reduces complexity, supports
cross-shard coordination, and optimizes resource usage.

The batcher will be updated in the next commits.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Add write_request_balancer tests

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
resource_mgmt: Add cloud topics

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
ct: Use memory_groups to limit CT memory

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

Lazin committed:
config: Add cfg for the write_reqeust_scheduler

The configs are 'development_cloud_topics_l0_cardinality_threshold',
'development_cloud_topics_l0_upload_interval' and
'development_cloud_topics_l0_data_threshold'.

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

asimms41 commented on src/v/config/configuration.cc:
```diff
@@ -4374,6 +4374,29 @@ configuration::configuration()
+  , development_cloud_topics_l0_data_threshold(
+      *this,
+      "development_cloud_topics_l0_data_threshold",
+      "Threshold for the L0 object size in cloud topics. When the amount of "
```
```suggestion
      "The size limit for the Level 0 object in cloud topics. When the amount of "
```

Lazin replied:
fixed

asimms41 commented on src/v/config/configuration.cc:
```diff
@@ -4374,6 +4374,29 @@ configuration::configuration()
+      "development_cloud_topics_l0_data_threshold",
+      "Threshold for the L0 object size in cloud topics. When the amount of "
+      "data on "
+      "a shard reaches the threshold the upload is triggered.",
```
```suggestion
      "a shard reaches this limit, an upload is triggered.",
```

asimms41 commented on src/v/config/configuration.cc:
```diff
@@ -4374,6 +4374,29 @@ configuration::configuration()
+  , development_cloud_topics_l0_cardinality_threshold(
+      *this,
+      "development_cloud_topics_l0_cardinality_threshold",
+      "Threshold for the L0 object cardinality in cloud topics. When the "
```
```suggestion
      "When the number of partitions in Level 0 of cloud topics reach this limit, an upload is triggered. "
```
Please check for accuracy.

Lazin replied:
tweaked this slightly to mention that the limit is per object

asimms41 commented on src/v/config/configuration.cc:
```diff
@@ -4374,6 +4374,29 @@ configuration::configuration()
+      "development_cloud_topics_l0_cardinality_threshold",
+      "Threshold for the L0 object cardinality in cloud topics. When the "
+      "number of "
+      "partitions waiting for the upload reaches the threshold the upload is "
```
```suggestion
```

asimms41 commented on src/v/config/configuration.cc:
```diff
@@ -4374,6 +4374,29 @@ configuration::configuration()
+      *this,
+      "development_cloud_topics_l0_cardinality_threshold",
+      "Threshold for the L0 object cardinality in cloud topics. When the "
+      "number of "
```
```suggestion
      ""
```

asimms41 reviewed (commented):
Some suggestions.

Lazin committed:
ct: Add write_request_balancer benchmark

Signed-off-by: Evgeny Lazin <4lazin@gmail.com>

rockwotj commented on src/v/config/configuration.cc:
```diff
@@ -4374,6 +4374,29 @@ configuration::configuration()
       "Enable cloud topics.",
       {.needs_restart = needs_restart::no, .visibility = visibility::user},
       false)
+  , development_cloud_topics_l0_data_threshold(
```
for these can we remove the `development` prefix? I don't think it buys anything and it will prevent churn later.

Lazin replied:
fixed

rockwotj commented on src/v/cloud_topics/level_zero/pipeline/write_pipeline.h:
```diff
@@ -73,7 +73,9 @@ class write_pipeline
-        write_requests_list pull_write_requests(size_t max_bytes);
+        write_requests_list pull_write_requests(
+          size_t max_bytes,
+          size_t max_requests = std::numeric_limits<size_t>::max());
```
It would be good to add to the method comment what max_bytes and max_requests are and how they work

Lazin replied:
done

rockwotj commented on src/v/cloud_topics/level_zero/write_request_scheduler/write_request_scheduler.cc:
```diff
@@ -0,0 +1,320 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
```
![what-year-is-it](https://github.com/user-attachments/assets/804a12ad-8bc5-4af1-8f86-5f84bac63bca)

Lazin replied:
the original implementation on which this PR is based was created in 2024, but I'll update

Lazin replied:
din't know you can post gifs here

rockwotj commented on src/v/cloud_topics/level_zero/pipeline/write_pipeline.h:
```diff
@@ -162,8 +164,10 @@ class write_pipeline
     /// Get write requests atomically.
     /// The total size of returned write requests and the stage to which they
     /// belong to should be specified.
```
Needs updating with the new parameter

Lazin replied:
done

rockwotj commented on src/v/cloud_topics/level_zero/write_request_scheduler/write_request_scheduler.h:
```diff
@@ -0,0 +1,96 @@
+
+namespace cloud_topics::l0 {
+
+/// The scheduler decides the which shard to use to handle write
+/// requests. It also batches requests.
+///
```
Could have some commentary on the algorithm used and the policy, as well that it's intended to maximize the object sizes uploaded, while respecting the latency goals specified in our configs. Maybe copying and pasting a lot of the commit message here would be good.

Lazin replied:
done

rockwotj commented on src/v/cloud_topics/level_zero/write_request_scheduler/write_request_scheduler.h:
```diff
@@ -0,0 +1,96 @@
+    ss::abort_source _as;
+    ss::gate _gate;
+
+    // This field is used in tests to limit the uploads to time based fallback
+    bool _test_only_disable_data_threshold{false};
+    // This field is used in tests to limit the uploads to data threshold
+    bool _test_only_disable_time_based_fallback{false};
```
shouldn't this be done by making the limits super high instead of a bool to disable?

Lazin replied:
I think this approach is a bit more readable. You see the configs being set in the beginning of the test and then you need to figure out what are they doing first. But with flags the intent is much more obvious. But I don't have stroing opinion.

rockwotj commented on src/v/cloud_topics/level_zero/write_request_scheduler/write_request_scheduler.cc:
```diff
@@ -0,0 +1,320 @@
+                     && total_bytes < _max_buffer_size()
+                   ? l0::request_processing_result::ignore_and_continue
+                   : l0::request_processing_result::ignore_and_stop;
+      });
```
nit: the `l0` namespace is not needed

Lazin replied:
fixed

rockwotj commented on src/v/cloud_topics/level_zero/write_request_scheduler/write_request_scheduler.cc:
```diff
@@ -0,0 +1,320 @@
+    co_return std::move(fp);
+}
+
+void write_request_scheduler::ack_write_response(
```
TBH I think it's simpler to just inline this method - it would remove the need for some of the comments.

Lazin replied:
I updated the call site so it makes more sense to use the function now.

rockwotj commented on src/v/cloud_topics/level_zero/write_request_scheduler/tests/write_request_scheduler_test.cc:
```diff
@@ -0,0 +1,335 @@
+TEST_F_CORO(write_request_balancer_fixture, time_based_fallback_test) {
+    // This test produces some batches and expects time based threshold
+    // mechanism to trigger the upload.
+    ASSERT_TRUE_CORO(ss::smp::count > 1);
```
you can set the number of cores used in the BUILD file BTW

Lazin replied:
it is set in the BUILD, this is just to simplify debugging if someone tries to run the test manually using wrong config

rockwotj reviewed (commented):
First round

rockwotj reviewed (commented):
LGTM, will let another person review and do the final approval.

paulohtb6 commented on src/v/config/configuration.cc:
```diff
@@ -4374,6 +4374,27 @@ configuration::configuration()
+  , cloud_topics_l0_data_threshold(
+      *this,
+      "cloud_topics_l0_data_threshold",
+      "The size limit for the Level 0 object size in cloud topics. When the "
```
hey @Lazin I saw that @asimms41 already gave you some suggestions on the topic. But after another read-through I have some other points. 


## Descriptions
I would try to make the descriptions consistent between themselves.

Here's some suggestions:

> cloud_topics_l0_data_threshold - The size threshold for Level 0 objects in cloud topics. When data on a shard reaches this threshold, Redpanda triggers an upload.


> cloud_topics_l0_upload_interval - The time interval threshold for Level 0 objects in cloud topics. When this interval elapses, Redpanda triggers an upload.


> cloud_topics_l0_cardinality_threshold - The partition count threshold for Level 0 objects in cloud topics. When the number of partitions in a Level 0 object reaches this threshold, Redpanda triggers an upload.

## Readability

Also, l0 has some readability changes (sometimes it looks like it's i o). I would suggest renaming the property itself to level_0 to avoid that. `cloud_topics_level_0_data_threshold` for example. 

## Is L0 cloud topics an user facing feature? 
btw, I have no idea what "level 0 cloud topics" stand for lol. Is this L0 like in cache layers? Is this part of a bigger feature?

Asking to see if we should go into detail on the description, or when we eventually ship this, we can add an entry to our glossary to explain it. If this is used as an internal-only thing, we can ignore that

Lazin replied:
I removed L0 from names and descriptions.

Lazin commented:
Rebased with dev and changed configuration names (l0 is replaced with produce)

Lazin commented on src/v/cloud_topics/level_zero/write_request_scheduler/write_request_scheduler.cc:
```diff
@@ -0,0 +1,341 @@
+
+namespace cloud_topics::l0 {
+
+namespace {
+// Copy extents and share the payload to use on another shard
+static inline serialized_chunk shallow_copy(serialized_chunk& chunk) {
+    serialized_chunk copy;
+    copy.extents = chunk.extents.copy();
+    copy.payload = chunk.payload.share(0, chunk.payload.size_bytes());
+    return copy;
+}
+} // namespace
```
It's a synchronous function, there is no lifetime considerations because it has no side effects. There is a doc string that describes the function.

dotnwat commented on src/v/cloud_topics/level_zero/write_request_scheduler/write_request_scheduler.cc:
```diff
@@ -0,0 +1,341 @@
+}
+
+ss::future<> write_request_scheduler::apply_time_based_fallback() {
+    // The time based fallback works in three stages:
+    // 1. Collect information about the requests from every shard.
+    // 2. Decide which shard should handle the requests.
+    // 3. Tell every shard to forward its requests to the target shard.
+    // The whole process is triggered on shard 0. The end to end latency
+    // of the process is relatively high (up to 1ms) but it's fine. The
+    // method is invoked periodically (e.g. every 500ms) so the
+    // additional latency is not significant. If we have very high tput
+    // on some shards they will be excluded from the time based fallback
+    // because they will trigger the data threshold based uploads.
```
it's fine for now, but this seems like it will have some non-optimal properties. the biggest one being that gang-scheduling across all the cores ignores arrival time. the observation is that the latency clock starts ticking when a request arrives, and only then is a cross-core action necessary. second, in the gang-scheduled approach:

> 3. Tell every shard to forward its requests to the target shard.

constraining the consolidation to a single node may be non-optimal.

Lazin replied:
it will be addressed in the next PR

Lazin replied:
This changes in the next PR slightly. Shard 0 collects information periodically and decides that it's time to trigger an upload. Then it invokes the target shard (the one which will be uploading the object). The target shard pulls data from other shards and uploads the object.

Lazin replied:
I'll write a doc with the high-level description of this process. The next PR omits few optimizations including the possibility for avoiding consolidation on one shard. It's easy to tweak the code so it'd schedule more than one shard to upload. The change is very trivial. But I suspect that it might not be necessary because in combination with the data threshold uploads shards will never accumulate enough data to justify such scheduling decision.

dotnwat replied:
thanks!

dotnwat reviewed (approved):
looks great.

it would be fantastic if we could discuss as a group high level designs before dropping code. for example, I think there are downsides to the gang scheduling used for the time-expiration based appraoch and it would save a lot of time to just form consensus on the approach before writing and reviewing code.
